presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java

```diff
@@ -118,37 +124,31 @@ public StripeReader(
         this.cacheable = requireNonNull(cacheable, "hiveFileContext is null");
     }
 
-    public Stripe readStripe(StripeInformation stripe, OrcAggregatedMemoryContext systemMemoryUsage)
+    public Stripe readStripe(StripeInformation stripe, OrcAggregatedMemoryContext systemMemoryUsage, List<Slice> decryptionKeyMetadata)
             throws IOException
     {
         StripeId stripeId = new StripeId(orcDataSource.getId(), stripe.getOffset());
 
         // read the stripe footer
         StripeFooter stripeFooter = readStripeFooter(stripeId, stripe, systemMemoryUsage);
-        List<ColumnEncoding> columnEncodings = stripeFooter.getColumnEncodings();
 
+        List<ColumnEncoding> columnEncodings = stripeFooter.getColumnEncodings();
         // get streams for selected columns
         Map<StreamId, Stream> streams = new HashMap<>();
         boolean hasRowGroupDictionary = false;
-        for (Stream stream : stripeFooter.getStreams()) {
-            if (includedOrcColumns.contains(stream.getColumn())) {
-                streams.put(new StreamId(stream), stream);
-
-                if (stream.getStreamKind() == StreamKind.IN_DICTIONARY) {
-                    ColumnEncoding columnEncoding = columnEncodings.get(stream.getColumn());
-
-                    if (columnEncoding.getColumnEncodingKind() == DICTIONARY) {
-                        hasRowGroupDictionary = true;
-                    }
-
-                    Optional<SortedMap<Integer, DwrfSequenceEncoding>> additionalSequenceEncodings = columnEncoding.getAdditionalSequenceEncodings();
-                    if (additionalSequenceEncodings.isPresent()
-                            && additionalSequenceEncodings.get().values().stream()
-                            .map(DwrfSequenceEncoding::getValueEncoding)
-                            .anyMatch(encoding -> encoding.getColumnEncodingKind() == DICTIONARY)) {
-                        hasRowGroupDictionary = true;
-                    }
-                }
+        hasRowGroupDictionary = addIncludedStreams(columnEncodings, stripeFooter.getStreams(), streams);
+
+        // if streams doesn't have all included orcColumns, some included columns are encrypted
+        // get encrypted streams
+        if (streams.size() < includedOrcColumns.size()) {
+            verify(decryptorProvider.isPresent(), "decryptorProvider is absent");
+            List<Slice> encryptedStripeEncryptionGroups = stripeFooter.getStripeEncryptionGroups();
+            for (int i = 0; i < encryptedStripeEncryptionGroups.size(); i++) {
+                DwrfDecryptor decryptor = decryptorProvider.get().createDecryptor(decryptionKeyMetadata.get(i));
```

Comment:
in theory, decrypter doesn't need to be recreated as long as key metadata doesn't change